arangod/Aql/DocumentProducingHelper.cpp

```diff
@@ -505,49 +517,55 @@ template<bool checkUniqueness, bool skip>
 IndexIterator::CoveringCallback aql::getCallback(
     DocumentProducingCallbackVariant::WithProjectionsCoveredByIndex,
     DocumentProducingFunctionContext& context) {
-  return [&context](LocalDocumentId token,
-                    IndexIteratorCoveringData& covering) {
-    TRI_ASSERT(context.getAllowCoveringIndexOptimization());
-    if constexpr (checkUniqueness) {
-      TRI_ASSERT(token.isSet());
-      if (!context.checkUniqueness(token)) {
-        // Document already found, skip it
-        return false;
-      }
-    }
+  if (context.hasFilter()) {
+    TRI_ASSERT(!context.getFilterProjections().empty());
+    TRI_ASSERT(context.getFilterProjections().usesCoveringIndex());
+    LOG_DEVEL << "############### filter can use index";
+  }
 
-    context.incrScanned();
+  return
+      [&context](LocalDocumentId token, IndexIteratorCoveringData& covering) {
+        TRI_ASSERT(context.getAllowCoveringIndexOptimization());
+        if constexpr (checkUniqueness) {
+          TRI_ASSERT(token.isSet());
+          if (!context.checkUniqueness(token)) {
+            // Document already found, skip it
+            return false;
+          }
+        }
 
-    // recycle our Builder object
-    VPackBuilder& objectBuilder = context.getBuilder();
-    if (!skip || context.hasFilter()) {
-      objectBuilder.clear();
-      objectBuilder.openObject(true);
+        context.incrScanned();
 
-      // projections from a covering index
-      context.getProjections().toVelocyPackFromIndex(objectBuilder, covering,
-                                                     context.getTrxPtr());
+        if (context.hasFilter() && !context.checkFilter(&covering)) {
+          context.incrFiltered();
+          return false;
+        }
 
-      objectBuilder.close();
+        // recycle our Builder object
+        VPackBuilder& objectBuilder = context.getBuilder();
+        if (!skip) {
+          objectBuilder.clear();
+          objectBuilder.openObject(true);
 
-      if (context.hasFilter() && !context.checkFilter(objectBuilder.slice())) {
-        context.incrFiltered();
-        return false;
-      }
-    }
-    if constexpr (!skip) {
-      InputAqlItemRow const& input = context.getInputRow();
-      OutputAqlItemRow& output = context.getOutputRow();
-      RegisterId registerId = context.getOutputRegister();
-      TRI_ASSERT(!output.isFull());
-      VPackSlice s = objectBuilder.slice();
-      output.moveValueInto(registerId, input, s);
-      TRI_ASSERT(output.produced());
-      output.advanceRow();
-    }
+          // projections from a covering index
+          context.getProjections().toVelocyPackFromIndex(
+              objectBuilder, covering, context.getTrxPtr());
 
-    return true;
-  };
+          objectBuilder.close();
+        }
+        if constexpr (!skip) {
```

Comment:
It looks like as if this condition (`! skip`) can be merged with the if above, which has the same condition.